qa/tasks/cephfs/caps_helper.py

```diff
@@ -27,7 +27,7 @@ def _gen_mon_cap_str(perm, fsname=None):
         return mon_cap
 
     if len(caps) == 1:
-        return _gen_mon_cap_str(caps[0])
+        return _gen_mon_cap_str(*caps[0])
```

Comment:
explain a bit in the commit message about the bug.

Reply:
I avoided it because this meant explaining a Python concept.

Reply:
I've elaborated on the Python concept in the commit message now.

Reply:
OK, thx. So, the caps didn't get generated as expected? Didn't that cause a test case failure?

Reply:
It does. I've spotted this once in integration test runs.

Reply:
Err... then how did those (buggy) set of changes get merged if the test case failed?

Reply:
@rishabh-d-dave 

```
<rishabh> overclk, which testcase? the bug that i am talking about was in helper methods in caps_helper.py
<overclk> rishabh: aren't the helpers used in tests?
<overclk> the test cases that use these helpers are likely to fail, isn't it?
```

Reply:
I tested few PRs in combination which was why this bug wasn't caught.

Reply:
could you point me to the run and the PRs list -- we should try to avoid such bugs getting merged.

Reply:
This is where the bug was introduced -

https://github.com/ceph/ceph/pull/50497/commits/969a93d0dcf1848623b0bd581b8e2e9a5a556839

Here's the teuthology runs -

https://github.com/ceph/ceph/pull/50497#issuecomment-1482296650

And the testing included PR #41779 along with PR #50497.

Reply:
> This is where the bug was introduced - [969a93d](https://github.com/ceph/ceph/commit/969a93d0dcf1848623b0bd581b8e2e9a5a556839).

I don't think so. This commit uses unpack_tuple which does not have the bug - callers for _gen_mds_cap_str() are:

```
    if len(caps) == 1:
       return _gen_mds_cap_str(caps[0])
```

and 

```
    for i, c in enumerate(caps):
       mds_cap +=  _gen_mds_cap_str(c)
       if i != len(caps) - 1:
           mds_cap += ', '
```

which seems fine.

> Here's the teuthology runs - [#50497 (comment)](https://github.com/ceph/ceph/pull/50497#issuecomment-1482296650) And the testing included PR #41779 along with PR #50497.

Isn't https://github.com/ceph/ceph/commit/87025d15858aa88b22afa9702511ccc7120f8b0b (PR #50664) where the bug got introduced? The test run for this is: http://pulpito.front.sepia.ceph.com/rishabh-2023-04-05_11:02:03-fs:functional-wip-rishabh-fs-auth-subcmd-distro-default-smithi/ - branch: wip-rishabh-fs-auth-subcmd which has the following code:

```
87025d15858a (Rishabh Dave 2023-03-29 20:48:53 +0530  71)     def _gen_mds_cap_str(perm, fsname=None, cephfs_mntpt='/'):
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  72)         mds_cap = f'allow {perm}'
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  73)         if fsname:
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  74)             mds_cap += f' fsname={fsname}'
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  75)         if cephfs_mntpt != '/':
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  76)             if cephfs_mntpt[0] == '/':
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  77)                 cephfs_mntpt = cephfs_mntpt[1:]
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  78)             mds_cap += f' path={cephfs_mntpt}'
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  79)         return mds_cap
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  80)
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  81)     if len(caps) == 1:
c8c29aaeb643 (Rishabh Dave 2023-04-07 20:43:46 +0530  82)         return _gen_mds_cap_str(*caps[0])
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  83)
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  84)     mds_cap = ''
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  85)     for i, c in enumerate(caps):
87025d15858a (Rishabh Dave 2023-03-29 20:48:53 +0530  86)         mds_cap +=  _gen_mds_cap_str(*c)
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  87)         if i != len(caps) - 1:
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  88)             mds_cap += ', '
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  89)
969a93d0dcf1 (Rishabh Dave 2023-03-13 16:16:13 +0530  90)     return mds_cap
```

where:

> c8c29aaeb643 (Rishabh Dave 2023-04-07 20:43:46 +0530  82)         return _gen_mds_cap_str(*caps[0])

is part of https://github.com/ceph/ceph/pull/41779/commits

and"

> 87025d15858a (Rishabh Dave 2023-03-29 20:48:53 +0530  86)         mds_cap +=  _gen_mds_cap_str(*c)

is a part of https://github.com/ceph/ceph/pull/50664/commits

It looks like the above PRs (#41779 and #50664) were included in the test branch, however, only #50664 was merged. @rishabh-d-dave ?

Reply:
> [c8c29aa](https://github.com/ceph/ceph/commit/c8c29aaeb643ca4dbf66ea20b23ed3b81e0bc08f) was not present on PR #41779. It was added to only recently when this PR was created.

```
commit 39ce51a53cf9c8602adc89354214146e6ac505c3 (HEAD -> wip-rishabh-fs-auth-subcmd, rishabh-d-dave/fs-auth-subcmd, ci/wip-rishabh-fs-auth-subcmd)
Author: Rishabh Dave <ridave@redhat.com>
Date:   Fri Apr 7 23:51:26 2023 +0530

    mon/AuthMonitor: log when parsing caps fails

    Signed-off-by: Rishabh Dave <ridave@redhat.com>

commit c8c29aaeb643ca4dbf66ea20b23ed3b81e0bc08f
Author: Rishabh Dave <ridave@redhat.com>
Date:   Fri Apr 7 20:43:46 2023 +0530

    qa/cephfs/caps_helper: fix a bug in methods that generate cap string

    Signed-off-by: Rishabh Dave <ridave@redhat.com>

```

The commit was a part of the test branch - might not have been in the PR, but surely in the test branch, which is a bit more concerning then (cherry-picked?).

Anyway - please add `Introduced-by: <commit>` in the commit message.

Reply:
Done - https://github.com/ceph/ceph/pull/51127/commits/64e3dd7e62cddb6364a95a8d47f42b8a6ee1ef4f.

As per our conversation on IRC I've moved this commit to separate PR.